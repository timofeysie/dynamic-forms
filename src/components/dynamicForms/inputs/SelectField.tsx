import React from 'react';

interface IProps {
    name: string;
    required: boolean;
    label: string;
    handleChange: any;
    selectValues: string [];
}

/** Issue # 5: Add gender */
const SelectField = ({
  name, required, handleChange, selectValues,
}: IProps) => (
  <div data-testid="selectid">
    {/* <label>{label}</label> TODO: decide on layout including labels */}
    <select
      name={name}
      required={required}
      onChange={handleChange}
    >
      <option value="">Select</option>
      {selectValues.map((values, i) => <option value={i} key={values + name}>{values}</option>)}
    </select>
  </div>
);

export default SelectField;
