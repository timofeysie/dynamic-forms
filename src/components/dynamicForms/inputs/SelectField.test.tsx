import React from 'react';
import { unmountComponentAtNode } from 'react-dom';
import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import SelectField from './SelectField';

// The following causes a linting error:
// TypeError: Cannot call method on null
// let container: any = null;
// was working previously...
let container = document.createElement('div');
const form = {
  name: 'gender',
  required: true,
  label: 'Select a gender',
  input_type: 'select',
  values: [
    'Male',
    'Female',
  ],
};

const onSubmitMock = jest.fn();
const setup = () => {
  const utils = render(<SelectField
    name={form.name}
    required={form.required}
    label={form.label}
    handleChange={onSubmitMock}
    selectValues={form.values}
    key={form.label + form.name}
  />);
  const input = utils.getByTestId('selectid') as HTMLInputElement;
  return {
    input,
    ...utils,
  };
};

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = document.createElement('div');
});

it('renders an input field', () => {
  const input = setup();
  expect(input).toBeDefined();
});

it('validates the input field', () => {
  const { input } = setup();
  form.values[0] = input.value;
  fireEvent.change(input);
  expect(input.value).toBe(form.values[0]);
});

it('matches the snapshot', () => {
  const tree = renderer.create(<SelectField
    name={form.name}
    required={form.required}
    label={form.label}
    handleChange={onSubmitMock}
    selectValues={form.values}
    key={form.label + form.name}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});
