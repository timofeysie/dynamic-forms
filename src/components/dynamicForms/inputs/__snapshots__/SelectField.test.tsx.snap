// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div
  data-testid="selectid"
>
  <select
    name="gender"
    onChange={[MockFunction]}
    required={true}
  >
    <option
      value=""
    >
      Select
    </option>
    <option
      value={0}
    />
    <option
      value={1}
    >
      Female
    </option>
  </select>
</div>
`;
